[train][checkpoint] Add validate_function and validate_config to ray.train.report

TimothySeah opened the pull request:
# Summary

The main change here is:
* Train workers report validation function + validation config
* Controller kicks off validation Ray task and associates its return value with the relevant checkpoint.
* Main controller step polls workers and validations, only finishing when both are done. 

In doing this, we are leveraging Ray Train's single controller architecture to provide a global understanding of training progress. This makes it easy to do stuff like early stopping in the future.

A few other notes:
* When validations fail, training moves on, but the controller prints the failure and stores it in the `Result`. I added a TODO to retry and time out in the future.
* We never delete pending checkpoints because they could end up being better than existing checkpoints. We may keep failed checkpoints depending on their initially reported metrics + the CheckpointConfig. Note that `Result.failed_validations` contains all the checkpoints that failed validations, which may or may not have been deleted. 
* I added a TODO to resume validations on `CheckpointManager` restoration; right now we simply won't rerun interrupted validations.
* I added a TODO to rate limit validations with a queue; right now we always kick off Ray tasks and rely on Ray Core scheduling to rate limit.

# API Examples

You can define a `validate_function` with `map_batches`

```
class Predictor:
    def __init__(self, checkpoint):
        self.model = create_model()
        with checkpoint.as_directory() as checkpoint_dir:
            model_state_dict = torch.load(os.path.join(checkpoint_dir, "model.pt"))
            self.model.load_state_dict(model_state_dict)
        self.model.cuda().eval()

    def __call__(self, batch):
        image = torch.as_tensor(batch['image'], dtype=torch.float32, device="cuda")
        label = torch.as_tensor(batch['label'], dtype=torch.float32, device="cuda")
        pred = self.model(image)
        return {'res': (pred.argmax(1) == label).cpu().numpy()}


def validate_with_map_batches(checkpoint, config):
    eval_res = config['dataset'].map_batches(
        Predictor,
        batch_size = 128,
        num_gpus=1,
        fn_constructor_kwargs = {'checkpoint': checkpoint},
        # guarantees worker on 2 actors because predictor is class.
        concurrency=2,
    )
    mean = eval_res.mean(['res'])
    return {
        'score': mean,
    }
```

or `TorchTrainer`. 

```
def eval_only_train_func(config_dict):
    # Load the checkpoint
    model = create_model()
    with config_dict['checkpoint'].as_directory() as checkpoint_dir:
        model_state_dict = torch.load(os.path.join(checkpoint_dir, "model.pt"))
        model.load_state_dict(model_state_dict)
    model.cuda().eval()

    # Set up metrics and data loaders
    criterion = CrossEntropyLoss()
    mean_valid_loss = torchmetrics.MeanMetric().cuda()
    test_data_shard = ray.train.get_dataset_shard("test")
    test_dataloader = test_data_shard.iter_torch_batches(batch_size=256)

    # Compute and report metric
    with torch.no_grad():
        for batch in test_dataloader:
            images, labels = batch['image'], batch['label']
            outputs = model(images)
            loss = criterion(outputs, labels)
            mean_valid_loss(loss)
    with tempfile.TemporaryDirectory() as temp_dir:
        ray.train.report(
            metrics={'score': mean_valid_loss.compute().item()}, 
            checkpoint=ray.train.Checkpoint.from_directory(temp_dir),
            checkpoint_upload_mode=CheckpointUploadMode.NO_UPLOAD,
        )

def validate_with_torch_trainer(checkpoint, config):
    trainer = ray.train.torch.TorchTrainer(
        eval_only_train_func,
        train_loop_config={'checkpoint': checkpoint},
        scaling_config=ray.train.ScalingConfig(num_workers=2, use_gpu=True),
        datasets={"test": config['dataset']},
    )
    result = trainer.fit()
    return result.metrics
```

Note the following about the `map_batches` and `TorchTrainer` methods:
* During the validation forward pass, the `map_batches` method's `__call__` function must move the batch to the device, but the `TorchTrainer` method's forward pass does not because the dataloader automatically does this.
* The `map_batches` method uses Ray Data's metric aggregation methods, whereas the `TorchTrainer` method uses Torch's.

Either way, you need to report with the validate function as follows:

```
def train_func():
    ...
    for epoch in epochs:
            if rank == 0:
                metrics = {"loss": loss.item(), "epoch": epoch}
                iteration_checkpoint_dir = os.path.join(temp_checkpoint_dir, f"epoch_{epoch}_rank_{ray.train.get_context().get_world_rank()}")
                os.makedirs(iteration_checkpoint_dir, exist_ok=True)
                torch.save(
                    model.module.state_dict(),
                    os.path.join(iteration_checkpoint_dir, "model.pt")
                )
                ray.train.report(
                    metrics,
                    checkpoint=ray.train.Checkpoint.from_directory(iteration_checkpoint_dir),
                    checkpoint_upload_mode=CheckpointUploadMode.ASYNC,
                    validate_function=validate_function,
                    validate_config={
                        'dataset': test_dataset,
                    },
                )
            else:
                ray.train.report({}, None)

trainer = ray.train.torch.TorchTrainer(
    train_func,
    scaling_config=scaling_config,
    datasets={"train": train_dataset},
    run_config=ray.train.RunConfig(
        storage_path="/mnt/cluster_storage",
    ),
)
result = trainer.fit()
```

Note that both methods pass the `test_dataset` directly as a global variable - let me know if there is a better way to do this. 

# Testing

I tried both API's above in an Anyscale workspace on 2 epochs:

The `map_batches` API:
* Took 4m42s

<img width="2150" height="51" alt="Screenshot 2025-09-15 at 6 53 34 PM" src="https://github.com/user-attachments/assets/a10cec87-52df-46ab-8759-c6f2d664f5bb" />

* The validations took 13.836704969406128s and 20.883531093597412s
* Total report blocking time was negligible (0.007100105285644531s)
* `result.best_checkpoints` at the end looks like `[(Checkpoint(filesystem=local, path=/mnt/cluster_storage/ray_train_run-2025-09-15_18-31-44/checkpoint_2025-09-15_18-34-05.089176), {'loss': 2.0584681034088135, 'epoch': 0, 'score': 0.2369}), (Checkpoint(filesystem=local, path=/mnt/cluster_storage/ray_train_run-2025-09-15_18-31-44/checkpoint_2025-09-15_18-36-07.816422), {'loss': 1.8418619632720947, 'epoch': 1, 'score': 0.3152})]`
* The additional e2e time incurred by validation, measured as (time `result.fit` returned) - (time train_func exited), was 23.8961102962s, which is equal to (time taken to set up last validation) + (time taken to perform last validation)

The `TorchTrainer` API: 
* Took 4m49s

<img width="1544" height="337" alt="Screenshot 2025-09-16 at 1 38 16 PM" src="https://github.com/user-attachments/assets/b73367c6-58d3-4a9f-99b3-e83ca6c1e46c" />

* Interestingly, according to my print statements, the validations took 22.306870937347412s and 22.350095510482788s, but according to the train runs above, they took 17s and 18s.
* Total report blocking time was negligible (0.006956338882446289s)
* `result.best_checkpoints` at the end looks like `[(Checkpoint(filesystem=local, path=/mnt/cluster_storage/06325ef3-fc29-48eb-bc21-f3fca71da238/checkpoint_2025-09-16_13-34-59.088135), {'loss': 1.7960023880004883, 'epoch': 0, 'score': 1.8543205261230469}), (Checkpoint(filesystem=local, path=/mnt/cluster_storage/06325ef3-fc29-48eb-bc21-f3fca71da238/checkpoint_2025-09-16_13-37-05.083466), {'loss': 1.615540862083435, 'epoch': 1, 'score': 1.6425774097442627})]`
* The additional e2e time incurred by validation was 26.8572409153s.

I also tested the `TorchTrainer` API with autoscaling enabled:

<img width="1514" height="146" alt="Screenshot 2025-09-16 at 2 14 41 PM" src="https://github.com/user-attachments/assets/71877c3d-b5d5-4178-979a-153e66ab8ba0" />

The results are basically the same as above, but interestingly:
* The first validation took much longer than the second one because it includes autoscaling time.
* The e2e time is around the same (in fact, it was actually shorter) because training continues during autoscaling and validation time is less than training time so validation can catch up to training.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce async checkpoint validation via validate_fn/validate_config in report(), add ValidationManager, and refactor reporting to a TrainingReport with pending-checkpoint handling.
> 
> - **API**:
>   - `ray.train.v2.api.train_fn_utils.report` now accepts `validate_fn` and `validate_config` to run async checkpoint validation.
> - **Execution/Controller**:
>   - New `ValidationManager` runs validation tasks, polls results, and updates checkpoint metrics.
>   - Wire `ValidationManager` into `TrainController` and `ReportCallbackHandler`.
> - **Reporting Pipeline**:
>   - Introduce `_TrainingReport` and `_ValidationSpec` to carry checkpoint, metrics, and validation spec end-to-end.
>   - Change `ReportCallback.after_report` signature to receive `training_report` plus per-worker metrics.
>   - `WorkerStatus` now holds `training_report` (renamed from `training_result`).
> - **Checkpointing**:
>   - `CheckpointManager` supports pending checkpoints and updates metrics post-validation; tie-break equal scores by report index; persist state before deletions.
>   - Enhanced `_insert_into_sorted_list` to accept typed items and optional tie-break map.
> - **User Callbacks**:
>   - `UserCallbackHandler.after_report` now sources checkpoint from `training_report`.
> - **Tests/Build**:
>   - Add tests for validation and pending-checkpoint logic; rename and add Bazel targets (`test_async_checkpointing_validation*`, `test_validation_manager`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6e1ddbc454b9fadb2fe7cd9c18d381fbabb0db0c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

TimothySeah committed:
[train][checkpoint] Add validate_function and validate_config to ray.train.report

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix straggling merge conflicts and clean up unit tests

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix validation closure

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix register_checkpoint ordering with validation

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
actually use validationerror

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix unit tests

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
guard against validate function invalid return type

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
remove unnecessary flag in unit test function

Signed-off-by: Timothy Seah <tseah@anyscale.com>

justinvyu reviewed (commented):
LFG

TimothySeah committed:
move TrainingReport/ValidationSpec to training_report.py

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
rename validate_function -> validate_fn

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
add TODOs

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
split out validationmanager and finish polling validations before controller shutdown

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -88,12 +109,26 @@ def __init__(
+        self._pending_training_results = {}
+
+        # Map from report number to ValidationInfo
+        self._failed_validations = OrderedDict()
```
done as discussed offline.

TimothySeah commented on python/ray/train/v2/_internal/execution/controller/controller.py:
```diff
@@ -402,8 +402,13 @@ async def _step(self) -> TrainControllerLoopIterationResult:
+            if (
+                worker_group_status.finished
+                and not worker_group_status.errors
+                and not self._checkpoint_manager.has_pending_validations()
```
moved to before_controller_shutdown

TimothySeah commented on python/ray/train/v2/_internal/execution/controller/controller.py:
```diff
@@ -402,8 +402,13 @@ async def _step(self) -> TrainControllerLoopIterationResult:
             return self._execute_resize_decision(controller_state.scaling_decision)
         elif isinstance(controller_state, RunningState):
             worker_group_status: WorkerGroupPollStatus = await self._poll_workers()
+            self._checkpoint_manager.poll_validations()
```
moved to `after_controller_state_update` as discussed offline

TimothySeah committed:
move to correct BUILD file

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix bazel target

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
fix unit test: None -> False

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/tests/test_checkpoint_manager.py:
```diff
@@ -166,13 +148,53 @@ async def test_before_init_train_context(tmp_path):
-    latest_checkpoint_result = _create_dummy_training_results(1, storage_context)[0]
-    checkpoint_manager.register_checkpoint(latest_checkpoint_result)
+    latest_checkpoint_result = create_dummy_training_results(1, storage_context)[0]
+    checkpoint_manager.register_checkpoint(latest_checkpoint_result, None)
```
done.

justinvyu commented on python/ray/train/v2/_internal/execution/controller/controller.py:
```diff
@@ -131,17 +134,22 @@ def __init__(
-            c for c in self._callbacks if isinstance(c, ControllerCallback)
-        ]
+        self._controller_callbacks = (
+            [self._scaling_policy]
+            + [c for c in self._callbacks if isinstance(c, ControllerCallback)]
+            + [self._validation_manager]
```
let's have validation manager come in front of potential user callbacks

```suggestion
            [self._scaling_policy, self._validation_manager]
            + [c for c in self._callbacks if isinstance(c, ControllerCallback)]
```

TimothySeah replied:
done.

justinvyu commented on python/ray/train/v2/_internal/execution/training_report.py:
```diff
@@ -0,0 +1,35 @@
+        return f"ValidationSpec(validate_fn={self.validate_fn}, validate_config={self.validate_config})"
+
+
+class _TrainingReport:
+    """A _TrainingResult reported by the user and a _ValidationSpec that describes how to validate it."""
```
Can we remove this extra _TrainingReport layer and just create a ValidationSpec that holds onto the _TrainingResult? I think it the checkpoint + (eager) metrics belong in the validation spec since you need to pass the checkpoint into the function.

TimothySeah replied:
To clarify, do you mean instead of

```
_TrainingReport
    _TrainingResult
    _ValidationSpec
        validate_fn
        validate_config
```

Do

```
_ValidationSpec
    _TrainingResult
    validate_fn
    validate_config 
```

?

In that case, would `execution/context.py` report a `_ValidationSpec`? I think this would be less intuitive because when you report, validation is optional. With the current setup:
* `execution/context.py` reports a `_TrainingReport` that contains a `_TrainingResult` (required) and _`ValidationSpec` (optional)
* All the `ReportCallbacks` are given eager metrics, checkpoint, and validation spec. `ValidationManager`’s after_report function can call `run_validate_fn` with `Checkpoint` and `ValidationSpec` since it doesn’t need the eager metrics. 

Lmk what you think.

TimothySeah replied:
discussed offline: this is fine for now but can have todo to have metrics consolidation function consolidates them in one trainingreport

TimothySeah replied:
added todo

justinvyu commented on python/ray/train/v2/api/validation_info.py:
```diff
@@ -0,0 +1,10 @@
+
+
+@dataclass
+class ValidationInfo:
```
```suggestion
class ValidationFailure:
```

TimothySeah replied:
done.

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py:
```diff
@@ -0,0 +1,108 @@
+    ):
+        self._checkpoint_manager = checkpoint_manager
+
+        # Map from checkpoint to validation task
+        self._pending_validations = {}
+
+        # Map from checkpoint to ValidationInfo
+        self._failed_validations = OrderedDict()
```
add todo for checkpointing validation manager state

TimothySeah replied:
done.

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py:
```diff
@@ -0,0 +1,108 @@
+                validation_spec, checkpoint
+            )
+
+    def poll_validations(self) -> bool:
```
```suggestion
    def _poll_validations(self) -> bool:
```

TimothySeah replied:
done.

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py:
```diff
@@ -0,0 +1,108 @@
+        """Poll pending validations, update checkpoint manager state, return whether pending validations remain."""
+        checkpoint_to_metrics = {}
+        checkpoints = list(self._pending_validations.keys())
+        for checkpoint in checkpoints:
+            validation_task = self._pending_validations[checkpoint]
+            done, _ = ray.wait([validation_task], timeout=0)
+            if done:
+                try:
+                    checkpoint_to_metrics[checkpoint] = ray.get(done[0])
+                except ray.exceptions.RayTaskError as e:
+                    checkpoint_to_metrics[checkpoint] = {}
+                    self._failed_validations[checkpoint] = ValidationInfo(
+                        checkpoint=checkpoint,
+                        validation_failed_error=ValidationFailedError(e.cause),
+                    )
+                    logger.exception(f"Validation failed for checkpoint {checkpoint}")
+                # TODO: retry validations and time out appropriately.
+                self._pending_validations.pop(checkpoint)
+        self._checkpoint_manager.update_checkpoints_with_metrics(checkpoint_to_metrics)
+        return bool(self._pending_validations)
```
do you think we should handle one finished validation at a time so that the controller doesn't get blocked for too long.

also, I think it's better to do ray.wait(many_tasks) instead of a bunch of `[ray.wait([task]) for task in tasks]`

TimothySeah replied:
Changed to:
* `ray.wait` for list of all pending tasks
* `ray.get` only the next (I use an OrderedDict now) task

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py:
```diff
@@ -0,0 +1,108 @@
+        if validation_spec and validation_spec.validate_fn:
+            # TODO: rate limit this by using a queue?
+            # TODO: figure out where to place run_validate_fn task:
+            # head node is faster but want to avoid putting too much there
+            self._pending_validations[checkpoint] = run_validate_fn.remote(
```
log an info message that we launched async validation

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py:
```diff
@@ -0,0 +1,108 @@
+            done, _ = ray.wait([validation_task], timeout=0)
+            if done:
+                try:
+                    checkpoint_to_metrics[checkpoint] = ray.get(done[0])
```
log that a validation finished + remaining pending

justinvyu replied:
Let's do a follow-up to improve logging. We can make it really clean and easy to parse from controller logs

TimothySeah replied:
Whenever a validation finishes, we print a log that looks like this

```
2025-09-20 16:08:28,614	INFO validation_manager.py:87 -- Finished async validation task for checkpoint(s) [Checkpoint(filesystem=local, path=/private/var/folders/3x/fk6f3pdx22bf6t9xx2l9tp1h0000gp/T/pytest-of-tseah/pytest-556/test_checkpoint_validation_man0/checkpoint_validation_management_experiment/checkpoint_0), Checkpoint(filesystem=local, path=/private/var/folders/3x/fk6f3pdx22bf6t9xx2l9tp1h0000gp/T/pytest-of-tseah/pytest-556/test_checkpoint_validation_man0/checkpoint_validation_management_experiment/checkpoint_1), Checkpoint(filesystem=local, path=/private/var/folders/3x/fk6f3pdx22bf6t9xx2l9tp1h0000gp/T/pytest-of-tseah/pytest-556/test_checkpoint_validation_man0/checkpoint_validation_management_experiment/checkpoint_3)]. Remaining pending validations for checkpoint(s): odict_values([Checkpoint(filesystem=local, path=/private/var/folders/3x/fk6f3pdx22bf6t9xx2l9tp1h0000gp/T/pytest-of-tseah/pytest-556/test_checkpoint_validation_man0/checkpoint_validation_management_experiment/checkpoint_2)])
```

lmk if this is fine or too verbose

justinvyu commented on python/ray/train/v2/tests/test_checkpoint_manager.py:
```diff
@@ -166,13 +148,53 @@ async def test_before_init_train_context(tmp_path):
+            training_results[1].checkpoint: {"score": 200},
+        }
+    )
+    assert checkpoint_manager._checkpoint_results == [
+        training_results[1],
+    ]
```
I kind of expected the better of the 2 "final" checkpoints to stick around as well.

Otherwise the unknown score always beats a finalized score.

Validations must be finalized anyways for the run to exit, so the end state will not violate the num_to_keep config.

justinvyu reviewed (commented):
Great! I'll do a more thorough pass on the tests in the next round.

TimothySeah committed:
address pr comments

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
make update_checkpoints_with_metrics more rigorous

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -118,14 +127,51 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
+            self._pending_training_results.pop(checkpoint)
+        self._save_state_and_delete_old_checkpoints()
+
+    def _save_state_and_delete_old_checkpoints(self):
```
> It can fail with a KeyError if a checkpoint is validated multiple times, as it's prematurely removed from _pending_training_results. 

> It also risks a ValueError when attempting to remove a checkpoint no longer in _checkpoint_results. 

Neither case should happen in practice but I guarded against them just in case.

> Additionally, modifying a checkpoint's sort key in-place while it's in the sorted _checkpoint_results list can corrupt the list's order under concurrent access.

The order will indeed be corrupted briefly but this is fine because there is no concurrent access.

justinvyu reviewed (commented):
Should be good after this!

TimothySeah committed:
remove validation failure logic

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
address some comments, mainly flatten TrainingReport + change after_report accordingly

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/report_handler.py:
```diff
@@ -26,8 +26,8 @@ def __init__(self, report_callbacks: List[ReportCallback]):
-        # A list of queues holding training results from workers.
-        self._training_result_queues: Optional[List[Deque[_TrainingResult]]] = None
+        # A list of queues holding training reports from workers.
+        self._training_reports_queues: Optional[List[Deque[_TrainingReport]]] = None
```
done.

TimothySeah committed:
address more comments

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/api/train_fn_utils.py:
```diff
@@ -19,6 +19,8 @@ def report(
     checkpoint_upload_mode: CheckpointUploadMode = CheckpointUploadMode.SYNC,
     delete_local_checkpoint_after_upload: Optional[bool] = None,
+    validate_fn: Optional[Callable[["Checkpoint", Optional[Dict]], Dict]] = None,
+    validate_config: Optional[Dict] = None,
```
done.

TimothySeah commented on python/ray/train/v2/tests/test_checkpoint_manager.py:
```diff
@@ -166,13 +148,114 @@ async def test_before_init_train_context(tmp_path):
+    training_results = create_dummy_training_results(
+        num_results=5, storage_context=storage_context
+    )
+    checkpoint_manager._pending_training_results[
+        training_results[0].checkpoint
+    ] = training_results[0]
+    checkpoint_manager._pending_training_results[
+        training_results[1].checkpoint
+    ] = training_results[1]
+    checkpoint_manager._pending_training_results[
+        training_results[2].checkpoint
+    ] = training_results[2]
+    checkpoint_manager._checkpoint_results.append(training_results[0])
+    checkpoint_manager._checkpoint_results.append(training_results[1])
+    checkpoint_manager._checkpoint_results.append(training_results[3])
```
I initially added this test just to appease Cursor (https://github.com/ray-project/ray/pull/56360#discussion_r2365868713), but after thinking about it more, I do think it makes sense to test that `update_checkpoints_with_metrics` works as expected even if `CheckpointManager`'s internal state is impossible given the current public methods. That way, if we change the internal state in the future, this test will still be valid.

That being said, overlaps are valid:
* result is in both `_checkpoint_results` and _pending_training_results`: `register_checkpoint` inserts the result into `_checkpoint_results` (with non-updated metrics), kicks off validation, and puts the result into `_pending_training_results`. After this happens the result is in both. 
* result is in `_checkpoint_results` only: this happens when there is no validation
* result is in `_pending_training_results` only: this can never happen given the current class interface (`register_checkpoint` and `update_checkpoints_with_metrics`), but I wanted to test this anyway for the reason above.

TimothySeah replied:
discussed offline - this is valid

TimothySeah committed:
add todo

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -118,14 +127,62 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
+                key=self._get_checkpoint_score,
+            )
+            self._pending_training_results.pop(checkpoint)
+        self._save_state_and_delete_old_checkpoints()
```
this should never happen so we raise an error now.

TimothySeah committed:
address more comments

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/v2/tests/test_checkpoint_manager.py:
```diff
@@ -166,13 +148,114 @@ async def test_before_init_train_context(tmp_path):
+            # ignored because not in _checkpoint_results
+            training_results[2].checkpoint: {"score": 50},
+            # ignored because not in _pending_training_results
+            training_results[4].checkpoint: {"score": 300},
```
after discussing offline i decided to raise after all

TimothySeah commented on python/ray/train/v2/tests/test_validation_manager.py:
```diff
@@ -0,0 +1,167 @@
+
+@unittest.mock.patch.object(
+    validation_manager.ValidationManager, "_poll_validations", autospec=True
+)
```
ended up just patching ray.wait only

TimothySeah commented on python/ray/train/v2/tests/test_async_checkpointing_validation.py:
```diff
@@ -217,6 +225,111 @@ def train_fn():
+        scaling_config=ScalingConfig(num_workers=2),
+    )
+    with pytest.raises(WorkerGroupError) as exc_info:
+        trainer.fit()
```
fixed.

TimothySeah commented on python/ray/train/v2/_internal/execution/context.py:
```diff
@@ -323,6 +336,14 @@ def report(
+                    checkpoint=checkpoint,
+                )
+            else:
+                validation_spec = None
```
Fixed. @justinvyu as part of this I decided to make `ValidationSpec` not include `Checkpoint` to:
* Avoid potential `Checkpoint` inconsistency
* Make a lot of the code cleaner. For example, if I were to implement `ValidationSpec` with `Checkpoint`, in `context.py` I would need to pass both validation_func and validation_config around until the persisted checkpoint is available.

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -118,14 +127,62 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
+
+    def _save_state_and_delete_old_checkpoints(self):
+        """Delete the old checkpoints."""
+        # Get checkpoints to delete
```
1) Initialized to set() 
2) I think this is a real issue - will take a look.

TimothySeah replied:
Done - see https://github.com/ray-project/ray/pull/56360#issuecomment-3340786281

TimothySeah committed:
use checkpoint report order as tiebreak

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented:
A few comments on https://github.com/ray-project/ray/pull/56360/commits/747e99b875b1af13a08f650d8b7924f4be277d03:

I considered the following alternate implementation methods but decided against them for various reasons:
* base `CheckpointManager` also updates `checkpoint_to_report_index`. I decided against this to minimize train v1 changes.
* Store report index inside trainingresult: this would be a large refactoring and expose unnecessary information to users

I reworked the `test_update_checkpoints_with_metrics`. It basically covered the same cases as `test_pending_checkpoint_management` except:
* trying to update checkpoint that is not pending: extracted into `test_pending_checkpoint_management_finalized_checkpoint`
* trying to update checkpoint that is not in checkpoint_results: already covered by `test_update_checkpoints_with_metrics_not_in_checkpoint_results`

I'm not worrying about restoring `_checkpoint_to_report_index` for now because:
* I already have a TODO to restore canceled validations
* Because we do not restore canceled validations, new validations will be associated with new reports, which will have report numbers.

TimothySeah commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -113,19 +128,68 @@ def register_checkpoint(self, checkpoint_result: _TrainingResult):
+                key=self._get_checkpoint_score,
+                checkpoint_to_report_index=self._checkpoint_to_report_index,
+            )
+            self._pending_training_results.pop(checkpoint)
```
non-issue because `update_checkpoints_with_metrics` is atomic

TimothySeah committed:
Merge remote-tracking branch 'upstream/master' into tseah/async-validate

justinvyu commented on python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py:
```diff
@@ -154,25 +161,25 @@ def update_checkpoints_with_metrics(
                 self._checkpoint_results,
                 checkpoint_result,
                 key=self._get_checkpoint_score,
+                checkpoint_to_report_index=self._checkpoint_to_report_index,
```
possible to just incorporate the checkpoint index as the default checkpoint score in `get_checkpoint_score`?

justinvyu replied:
oh I see your comment. you wanted to minimize train v1 changes?

justinvyu reviewed (approved):
🚢

TimothySeah committed:
change _insert_into_sorted_list typing

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented on python/ray/train/_internal/checkpoint_manager.py:
```diff
@@ -33,10 +45,17 @@ def _insert_into_sorted_list(list: List[Any], item: Any, key: Callable[[Any], An
+            list_item_key == item_key
+            and checkpoint_to_report_index.get(list[i].checkpoint, 0)
+            > checkpoint_to_report_index.get(item.checkpoint, 0)
+        ):
```
done.

TimothySeah committed:
pydoclint

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah committed:
remove example to fix doctest

Signed-off-by: Timothy Seah <tseah@anyscale.com>

TimothySeah commented:
> > If you use the TorchTrainer way, you also need to pass the experiment_dir_name to comply with the requirement that each TorchTrainer manages checkpoints from the same experiment directory.
> 
> Why is using the same experiment directory a requirement here?
> 
> If you reuse the same storage_path+name, then you'll end up "restoring" the previous run: https://docs.ray.io/en/latest/train/user-guides/fault-tolerance.html#job-driver-fault-tolerance This is probably why you were running into the "checkpoint directory" matching prefix assertion error. We don't need to "restore" the training run while launching the validation run.
> 
> The validation run shouldn't actually need to save any files or record checkpoints, so the storage path will be mostly unused.
> 
> I think we should recommend something like this:
> 
> ```python
> def eval_only_train_func(config_dict):
>     # ...
> 
>     # ray.train.report(metrics, checkpoint, NO_UPLOAD)
>     # !! The previous usage of ray.train.report w/ NO_UPLOAD is a bit confusing,
>     # since you're essentially checkpointing within a validation loop.
>     # The only reason why we needed the dummy checkpoint previously was so
>     # that we could get these `metrics` out of the training function.
>     metrics = {'score': mean_valid_loss.compute().item()}
>     return metrics
> 
> def validate_with_torch_trainer(checkpoint, config):
>     trainer = ray.train.torch.TorchTrainer(
>         eval_only_train_func,
>         train_loop_config={'checkpoint': checkpoint},
>         scaling_config=ray.train.ScalingConfig(num_workers=2, use_gpu=True),
>         # !! Just leave the default auto-generated UUID run name.
>         run_config=ray.train.RunConfig(storage_path="/mnt/cluster_storage"),
>         datasets={"test": config['dataset']},
>     )
>     result = trainer.fit()
>     return result.return_values[0]
> ```
> 
> Note that `result.return_values` is a new API but I think it's generally useful (user have asked for something similar) and avoids the need to attach dummy checkpoints during validation. We can recommend the dummy checkpoint workaround for now though, while deciding on the return values API.

Thanks, good catch! I agree that `return_values` is a better API. Like you pointed out, I was only calling `report` in `eval_only_train_func` as a workaround. I initially required both `TorchTrainers` to share the experiment directory so that the eval worker can report the checkpoint being validated, but after thinking about it more, I realize that we can just report a dummy checkpoint instead, which removes that requirement. I've also update the example in the PR description accordingly.
